src/OpenApi/test/Services/OpenApiDocumentProviderTests.cs

```diff
@@ -0,0 +1,73 @@
+using Microsoft.AspNetCore.OpenApi;
+using Microsoft.Extensions.ApiDescriptions;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.OpenApi.Models;
+using Microsoft.OpenApi.Readers;
+using static Microsoft.AspNetCore.OpenApi.Tests.OpenApiOperationGeneratorTests;
+
+public class OpenApiDocumentProviderTests
+{
+    [Fact]
+    public async Task GenerateAsync_ReturnsDocument()
+    {
+        // Arrange
+        var documentName = "v1";
+        var hostEnvironment = new HostEnvironment() { ApplicationName = nameof(OpenApiDocumentProviderTests) };
+        var documentService = new OpenApiDocumentService(hostEnvironment);
+        var serviceCollection = new ServiceCollection();
+        serviceCollection
+            .AddOpenApi(documentName)
+            .AddSingleton<IHostEnvironment>(hostEnvironment)
+            .AddSingleton(documentService);
+        var serviceProvider = serviceCollection.BuildServiceProvider();
+        var documentProvider = new OpenApiDocumentProvider(serviceProvider);
+        var stringWriter = new StringWriter();
+
+        // Act
+        await documentProvider.GenerateAsync(documentName, stringWriter);
+
+        // Assert
+        ValidateOpenApiDocument(stringWriter, document =>
+        {
+            Assert.Equal(hostEnvironment.ApplicationName, document.Info.Title);
+            Assert.Equal("1.0.0", document.Info.Version);
+        });
+    }
+
+    [Fact]
+    public void GetDocumentNames_ReturnsAllRegisteredDocumentName()
+    {
+        // Arrange
+        var hostEnvironment = new HostEnvironment() { ApplicationName = nameof(OpenApiDocumentProviderTests) };
+        var documentService = new OpenApiDocumentService(hostEnvironment);
+        var serviceCollection = new ServiceCollection();
+        serviceCollection
+            .AddOpenApi("v2")
```

Comment:
What happens if I call `.AddOpenApi("foo").AddOpenApi("foo")`?  Is it worth testing?

Reply:
We're used keyed DI under the hood so the last registration is used. For the exact scenario you presented, you'll get virtually the same behavior. If you did something like this though:

```sharp
services
	.AddOpenApi("v1")
	.AddOpenApi("v1", options => options.OpenApiVersion = OpenApiSpecVersion.OpenApi2_0);
```

Then you'll get a document serialized using the OpenApi v2 since the last registration is favored.

Reply:
I'm pretty sure you'll get the document serialized using OpenApio v2 in the given the opposite order too, right?

```csharp
services
	.AddOpenApi("v1", options => options.OpenApiVersion = OpenApiSpecVersion.OpenApi2_0)
	.AddOpenApi("v1");
```

Reply:
Sounds like we need a test. 😄

Reply:
@halter73 Yep, you're right. The test I added [here](https://github.com/dotnet/aspnetcore/blob/b2984a639b67543f9ef45b0d7a59845cb4450092/src/OpenApi/test/Extensions/OpenApiServiceCollectionExtensionsTests.cs#L144) didn't cover this scenario since I was overriding the value of `OpenApiVersion` in the second call.

Digging deeper into the Options code, it looks like each registered `IConfigureOptions` instance is called for the same named option ([ref](https://source.dot.net/#Microsoft.Extensions.Options/OptionsFactory.cs,58)), so what you effectively get is a union of all the configured options.

Perhaps this is "fine" behavior. Thoughts?

Reply:
The reason you might want to add a document twice is to modify the configuration of a document added elsewhere in the code?  i.e. there's a reason to not just throw when this happens (which I'm sure is unidiomatic in aspnetcore).

Reply:
I jotted this as an implementation consideration and will solicit feedback on the API issue associated with this.